[CompilerPerf] Re-implement memory-mapped file reader for #1094

jack-pappas opened the pull request:
I've re-implemented the memory-mapped file reader (inheriting from BinaryFile) so it uses the APIs in System.IO.MemoryMappedFiles instead of win32 API calls. This allows the more-efficient reader to be used on all platforms.

See issue #1094 for additional background.

KevinRansom commented:
@jack-pappas this looks very nice.

saul commented on src/absil/ilread.fs:
```diff
@@ -105,113 +106,79 @@ type BinaryFile() =
+            // to be able to read the file while we're reading it, instead of opening it for exclusive access.
+            // Don't dispose the FileStream (e.g., with a 'use' binding); the MemoryMappedFile takes ownership
+            // of the stream once it's passed in and will take care of disposing it.
+            // TODO: Should we use FileShare.Read here instead of FileShare.ReadWrite (used to maintain the same functionality
```
If you have a file open for read, another process can't open it for write. We don't need to open a file for read/write just to make sure another process doesn't modify the file underneath us.

dsyme replied:
So this should be ``FileShare.Read``?

jack-pappas replied:
Yes, I think it should be changed to ``FileShare.Read`` - ``FileShare.ReadWrite`` means other processes are able to open the file for both reading and writing, which I don't think we want.

jack-pappas commented:
@KevinRansom Looks like CI part 3 is failing, is that the CoreCLR builds? ``MemoryMappedFile`` lives in _System.Core_ for .NET Framework, but has it's own package (_System.IO.MemoryMappedFiles_) for CoreCLR, so we'll have to add a reference to that package. Do you know which project/config files I need to add that reference to? I haven't dug into the CoreCLR build logic for the compiler yet (although I can probably brute-force if needed 🤔).

KevinRansom commented:
@jack-pappas I will take care of that.

KevinRansom commented:
@jack-pappas 

I added :     "System.IO.MemoryMappedFiles":"4.4.0-beta-24913-02",
to : C:\microsoft\visualfsharp\src\fsharp\FSharp.Compiler\project.json\\

and got the following build errors, so it would seem that MemoryMappedFiles got some API reworking:

````
"c:\microsoft\visualfsharp\build-everything.proj" (default target) (1) ->
"c:\microsoft\visualfsharp\src\fsharp\Fsc\Fsc.fsproj" (Build target) (6) ->
"c:\microsoft\visualfsharp\src\fsharp\FSharp.Compiler\FSharp.Compiler.fsproj" (default target) (4:3) ->
(CoreCompile target) ->
  c:\microsoft\visualfsharp\src\absil\ilread.fs(125,13): error FS0505: The member or object constructor 'CreateFromFile' does not take 7 argument(s). An overload was found taki
ng 1 arguments. [c:\microsoft\visualfsharp\src\fsharp\FSharp.Compiler\FSharp.Compiler.fsproj]
  c:\microsoft\visualfsharp\src\absil\ilread.fs(143,27): error FS0039: The field, constructor or member 'ReadArray' is not defined. Maybe you want one of the following:   Read
Char   ReadDecimal [c:\microsoft\visualfsharp\src\fsharp\FSharp.Compiler\FSharp.Compiler.fsproj]
  c:\microsoft\visualfsharp\src\absil\ilread.fs(166,60): error FS0039: The field, constructor or member 'PrepareConstrainedRegions' is not defined. [c:\microsoft\visualfsharp\s
rc\fsharp\FSharp.Compiler\FSharp.Compiler.fsproj]

````

dsyme commented:
@jack-pappas @KevinRansom We need to compiler-perf check this.  I've added the label and will run some tests

jack-pappas commented:
@KevinRansom We may also need to add a reference to ``System.IO.UnmanagedMemoryStream``. ``MemoryMappedViewAccessor`` derives from ``System.IO.UnmanagedMemoryAccessor`` which is defined in that package; the ``System.IO.MemoryMappedFiles`` package references that package but perhaps a direct reference would fix the issue. I tried looking for the code for that base class in ``dotnet/corefx`` to see if the API has actually changed but couldn't find it anywhere...

I've added fixes (I think) for the other errors you saw when you tried to compile on coreclr.

KevinRansom commented:
@jack-pappas 
source is here:   https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/IO/UnmanagedMemoryAccessor.cs

jack-pappas commented:
@KevinRansom I see the ``ReadArray<T>`` method is there; any idea why it's not working on coreclr then?

KevinRansom commented:
@jack-pappas , I will take a look.

enricosada commented:
> I added : "System.IO.MemoryMappedFiles":"4.4.0-beta-24913-02",
> to : C:\microsoft\visualfsharp\src\fsharp\FSharp.Compiler\project.json\

@KevinRansom can we use the release version of the package? that is prerelease in a dev feed

- [4.0.0](https://www.nuget.org/packages/System.IO.MemoryMappedFiles/4.0.0) for .net core 1.0
- [4.3.0](https://www.nuget.org/packages/System.IO.MemoryMappedFiles/4.3.0) for .net core 1.1

jack-pappas commented:
@KevinRansom I tried adding the package reference to the ``project.json``, but the CI build (part 3) still failed. I then tried compiling this PR on my Windows machine and got the same error; I examined the ``System.IO.UnmanagedMemoryStream`` assembly from the package's ``netstandard13`` folder using ILSpy, and indeed the ``ReadArray<T>`` and ``Read<T>`` methods are missing from the ``UnmanagedMemoryAccessor`` class:

![pr-2528-missing-readarray](https://cloud.githubusercontent.com/assets/477287/24085892/4fb7b64a-0cda-11e7-9eee-74feb0d94b61.png)

jack-pappas commented:
@KevinRansom Any luck figuring out why the ``ReadArray<T>`` method is missing from the package? If not, I'll file an issue on corefx to investigate.

jack-pappas commented:
@KevinRansom I opened an issue on dotnet/corefx (linked above), it turns out the [``Read<T>``](https://apisof.net/catalog/System.IO.UnmanagedMemoryAccessor.Read%3CT%3E(Int64,T)) and [``ReadArray<T>``](https://apisof.net/catalog/System.IO.UnmanagedMemoryAccessor.ReadArray%3CT%3E(Int64,T(),Int32,Int32)) methods were removed from the .NET Core 1.x / ``netstandard`` 1.x releases, but have been added back for ``netstandard2.0``.

Given this code (in the PR) is only used within the compiler, are there any qualms with having the F# compiler require ``netstandard2.0`` when that's released? I understand we'll want to keep targeting older ``netstandard`` versions for ``FSharp.Core`` for backwards-compatibility, but it seems fine to require developers to have a fairly recent .NET Core version to run the compiler itself. In fact, if we target ``netstandard2.0`` for the compiler, we might not need different builds for .NET Framework and .NET Core, since there shouldn't be any differences in APIs (that we use) at that point.

cartermp commented:
@jack-pappas Personally I think it makes sense to have the F# compiler itself target `netcoreapp2.0` and take dependencies on `netstandard2.0`.  As you said, this can simplify many things for us.  I'll defer to @KevinRansom on this, but I certainly would support that.

KevinRansom commented:
@jack-pappas We think we will need to do a bunch of mucking about with the build to target netstandard 2.0  The reason is we use an old (about November) version of dotnet cli for coreclr build and testing.  The buildtools we require are also bound to an old version.  And so once that change is made I think targeting dotnet 2.0 will be fine, until then we're kind of in a bind.

Kevin

enricosada commented:
> Personally I think it makes sense to have the F# compiler itself target netcoreapp2.0 and take dependencies on netstandard2.0. As you said, this can simplify many things for us

Please is possibile to not depends on netstandard 2.0 only api or .net core 2.0 too son? or at least continue to support netstandard1.6 for a bit? If we need a patch and update of fsc, we can release it. 
.NET Core Sdk 2.0 is expected Q3, is far.

I think we need at least 1 release of fsc compatible with .net core 1.1, because that's required for some os, like ubuntu 16.10 (ref https://github.com/dotnet/netcorecli-fsc/issues/91 )

when sdk 1.0 rtm will be used in repo, will be easier to cross target (netstandard1.6, netstandard2.0), so for fsproj will be ok.

Not for ever, just in short/medium timeframe

dsyme commented:
@jack-pappas Can you resolve conflicts with master please? Thanks

jack-pappas commented:
@dsyme Sure -- it's done.

jack-pappas commented:
@KevinRansom I was thinking -- what about if I keep the old (P/Invoke-based) reader code using an ifdef so it's only compiled in for CoreCLR and otherwise use the new code? That'd let us get this PR merged and allow the memory-mapped file reader to be used on Mono for now; later on, when we want to target ``netappcore2.0`` or ``netstandard2.0`` or whatever, all we'll need to do is just remove the P/Invoke-based code.

enricosada commented:
@jack-pappas why .netcore will be special? if the `System.IO.MemoryMappedFiles` used is the one released (4.0.0 or 4.3.0) is ok ihmo to use it. 

Just not use netstandard2.0 at all, ihmo, until build script support so. And also there we should provide netstandard1.0/1.1 compilation anyway for medium term (with nre sdk will not be an issue at all support both).

jack-pappas commented:
@enricosada .NET Core needs special treatment at this time because ``netstandard1.6`` is missing a couple of methods we need to implement this properly. (See the discussion above for details.)

I agree that we need to wait to use ``netstandard2.0`` / ``netcoreapp2.0`` until a later time, which is specifically why we'd need to special-case .NET Core support here to continue using the old P/Invoke-based code.

enricosada commented:
There isn't not a try catch to fallback to normal without memory mapped?
Are we really sure is using pinvoke on netcoreapp? Becuade on unix/Mac is not doing so, so maybe we can just remove pinvoke and use the fallback?

enricosada commented:
Need to check if pinvoke is called at least on windows for fsc built for netcoreapp1.0

7sharp9 commented:
@jack-pappas Did you see improvements with this?

dsyme commented:
Some pref results are here: https://github.com/Microsoft/visualfsharp/blob/master/tests/scripts/compiler-perf-results.txt#L88. They don't reveal much difference

I think we may want to take this change just for code sanity.  And possibly it improves perf on Mono

jack-pappas commented:
Are we building the compiler against ``netstandard2.0`` or `netcoreapp2.0`` now? If so I'll fix this to build against that TFM. Otherwise, I'll try @enricosada's approach, so we can get this committed.

@7sharp9 No, I haven't seen perf improvements, but I didn't benchmark it either. I'm making this change primarily as a simplification/consolidation.

KevinRansom commented:
Oh my gosh this is quite out of date ...  I will work on getting it ready to merge.

jack-pappas commented:
I can rebase the changes into the last master and update the PR if that’s helpful.

KevinRansom commented:
Sure that saves me doing it :-)

Thanks

jack-pappas commented:
@KevinRansom I've rebased the changes for this PR onto the latest from ``master`` and updated the PR.

I haven't looked at this in a while though -- what did we ever decide to do for a version of ``netstandard`` / ``netcoreapp`` to target? The Ubuntu build failed due to not being able to find the ``System.IO.MemoryMappedFile`` package; do I need to also update (e.g.) project files to pull that in?

KevinRansom commented:
@jack-pappas I will take a look at it.

jack-pappas commented:
@KevinRansom Did you have a chance to take a look at the remaining build issues here?

KevinRansom commented:
Not yet

KevinRansom commented:
@dotnet-bot test this please

cartermp commented:
@dotnet-bot test this please

forki commented:
@dotnet-bot test this please

KevinRansom commented:
@dsyme, @jack-pappas 

Is this still of interest, or can I close it?

Thanks

Kevin

dsyme commented:
@KevinRansom Let's leave it open, although the most likely result is that steal parts for #4528

KevinRansom commented:
@dsyme yep sounds like a plan.
